File: src/test/java/workshop/part2/supervisor/VettingSupervisorTest.java
package workshop.part2.supervisor;

import java.util.concurrent.TimeUnit;

import akka.actor.Props;
import akka.actor.SupervisorStrategy;
import akka.testkit.TestActorRef;
import akka.testkit.TestProbe;
import org.junit.Test;
import scala.concurrent.duration.Duration;
import workshop.common.ad.Ad;
import workshop.part1.AkkaTest;
import workshop.part1.Verdict.VerdictType;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.Is.is;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class VettingSupervisorTest extends AkkaTest {

    @Test
    public void createsVettingActorWhenAdIsReceived () {
        VettingActorFactory vettingActorFactory = mock(VettingActorFactory.class);

        TestProbe vettingActor = TestProbe.apply(system);
        when(vettingActorFactory.create(any()))
            .thenReturn(vettingActor.ref());

        Ad ad = createAd(123);
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref());

        assertThat(vettingActor.expectMsgClass(Duration.create(0, TimeUnit.SECONDS), Ad.class), is(ad));
    }

    @Test
    public void spoofsSenderWhenCreatingVettingActor() {
        VettingActorFactory vettingActorFactory = mock(VettingActorFactory.class);

        TestProbe vettingActor = TestProbe.apply(system);
        when(vettingActorFactory.create(any()))
            .thenReturn(vettingActor.ref());

        Ad ad = createAd(123);
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref());

        vettingActor.expectMsgClass(Duration.create(0, TimeUnit.SECONDS), Ad.class);
        vettingActor.reply(VerdictType.GOOD);

        assertThat(sender.expectMsgClass(Duration.create(0, TimeUnit.SECONDS), VerdictType.class), is(VerdictType.GOOD));
    }

    @Test
    public void resumesVettingActorWhenUserNotFoundException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new UserNotFoundException("user not found test exception"));

        assertThat(strategy, is(SupervisorStrategy.resume()));
    }

    @Test
    public void restartsVettingActorWhenNullPointerException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new NullPointerException("null pointer test exception"));

        assertThat(strategy, is(SupervisorStrategy.restart()));
    }

    @Test
    public void escalatesExceptionWhenVettingActorFailsWithOtherException() {
        SupervisorStrategy.Directive strategy = createVettingSupervisor(mock(VettingActorFactory.class))
            .underlyingActor()
            .supervisorStrategy()
            .decider()
            .apply(new RuntimeException("other exception"));

        assertThat(strategy, is(SupervisorStrategy.escalate()));
    }

    private TestActorRef<VettingSupervisor> createVettingSupervisor(VettingActorFactory vettingActorFactory) {
        return TestActorRef.create(system, Props.create(VettingSupervisor.class, () -> new VettingSupervisor(vettingActorFactory)));
    }

    private Ad createAd(int userId) {
        return new Ad(userId, "ad title", "description");
    }
}

File: src/test/kotlin/workshop/part2/supervisor/VettingSupervisorTest.kt
package workshop.part2.supervisor

import akka.actor.ActorRef
import akka.actor.ActorRef.noSender
import akka.actor.Props
import akka.actor.SupervisorStrategy
import akka.testkit.JavaTestKit.duration
import akka.testkit.TestActor
import akka.testkit.TestActorRef
import akka.testkit.TestKit
import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.whenever
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.mock
import org.mockito.runners.MockitoJUnitRunner
import workshop.common.ad.Ad
import workshop.common.userservice.UserCriminalRecord
import workshop.part1.AkkaTest
import workshop.part1.Verdict
import workshop.part2.FraudWordActor.ExamineWordsResult
import workshop.part2.UserActor.CheckUserResult

@RunWith(MockitoJUnitRunner::class)
class VettingSupervisorTest : AkkaTest() {

    @Test
    fun createsVettingActorWhenAdIsReceived() {
        val vettingActorFactory = mock(VettingActorFactory::class.java)

        whenever(vettingActorFactory.create(any()))
                .thenReturn(sender.ref())

        val ad = createAd(123)
        createVettingSupervisor(vettingActorFactory).tell(ad, sender.ref())

        verify(vettingActorFactory).create(any())
        assertThat(sender.expectMsgClass(Ad::class.java), equalTo(ad))
    }

    @Test
    fun handlesVettingRequestsInParallel() {
        object : TestKit(system) {
            init {
                val userActor = TestKit(system)
                userActor.setAutoPilot(object : TestActor.AutoPilot() {
                    override fun run(sender: ActorRef, msg: Any): TestActor.AutoPilot {
                        sender.tell(CheckUserResult(UserCriminalRecord.GOOD), noSender())
                        return keepRunning()
                    }
                })

                val fraudWordActor = TestKit(system)
                fraudWordActor.setAutoPilot(object : TestActor.AutoPilot() {
                    override fun run(sender: ActorRef, msg: Any): TestActor.AutoPilot {
                        sender.tell(ExamineWordsResult(emptyList()), noSender())
                        return keepRunning()
                    }
                })

                val vettingActorFactory = VettingActorFactory(userActor.testActor(), fraudWordActor.testActor())
                val vettingActor = system.actorOf(Props.create(VettingSupervisor::class.java) { VettingSupervisor(vettingActorFactory) })

                vettingActor.tell(createAd(1), testActor())
                vettingActor.tell(createAd(2), testActor())

                within<Any>(duration("3 seconds")) {

                    // Ideally we should different responses
                    assertThat(expectMsgClass(Verdict::class.java), equalTo((Verdict.GOOD)))
                    assertThat(expectMsgClass(Verdict::class.java), equalTo((Verdict.GOOD)))
                }
            }
        }
    }

    @Test
    fun restartsVettingActorWhenNullPointerException() {
        val strategy = createVettingSupervisor(mock(VettingActorFactory::class.java))
                .underlyingActor()
                .supervisorStrategy()
                .decider()
                .apply(NullPointerException("test exception"))

        assertThat(strategy, equalTo(SupervisorStrategy.restart() as SupervisorStrategy.Directive))
    }

    @Test
    fun escalatesExceptionWhenVettingActorFailsWithOtherExceptionThanNullPointerException() {
        val strategy = createVettingSupervisor(mock(VettingActorFactory::class.java))
                .underlyingActor()
                .supervisorStrategy()
                .decider()
                .apply(RuntimeException("test exception"))

        assertThat(strategy, equalTo(SupervisorStrategy.escalate() as SupervisorStrategy.Directive))
    }

    private fun createVettingSupervisor(vettingActorFactory: VettingActorFactory): TestActorRef<VettingSupervisor> {
        return TestActorRef.create(system, Props.create(VettingSupervisor::class.java) { VettingSupervisor(vettingActorFactory) })
    }

    private fun createAd(userId: Int): Ad {
        return Ad(userId, "ad title", "description")
    }
}